servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectLBHttpConnectionFactory.java

```diff
@@ -0,0 +1,174 @@
+            final ConnectionFactoryFilter<ResolvedAddress, FilterableStreamingHttpConnection> connectionFactoryFilter,
+            final ProtocolBinding protocolBinding) {
+        super(config, executionContext, version -> reqRespFactory, connectStrategy, connectionFactoryFilter,
+                connectionFilterFunction, protocolBinding);
+        requireNonNull(config.h1Config(), "H1ProtocolConfig is required");
+        assert config.connectAddress() != null;
+        connectAddress = config.connectAddress().toString();
+    }
+
+    @Override
+    Single<FilterableStreamingHttpConnection> newFilterableConnection(final ResolvedAddress resolvedAddress,
+                                                                      final TransportObserver observer) {
+        assert config.h1Config() != null;
+        return buildStreaming(executionContext, resolvedAddress, config, observer)
+                .map(c -> new PipelinedStreamingHttpConnection(c, config.h1Config(),
+                        reqRespFactoryFunc.apply(HTTP_1_1), config.allowDropTrailersReadFromTransport()))
+                .flatMap(this::processConnect);
+    }
+
+    // Visible for testing
+    Single<FilterableStreamingHttpConnection> processConnect(final NettyFilterableStreamingHttpConnection c) {
+        try {
+            // Send CONNECT request: https://datatracker.ietf.org/doc/html/rfc9110#section-9.3.6
+            // Host header value must be equal to CONNECT request target, see
+            // https://github.com/haproxy/haproxy/issues/1159
+            // https://datatracker.ietf.org/doc/html/rfc7230#section-5.4:
+            //   If the target URI includes an authority component, then a client MUST send a field-value
+            //   for Host that is identical to that authority component
+            final StreamingHttpRequest request = c.connect(connectAddress).setHeader(HOST, connectAddress);
+            // No need to offload because there is no user code involved
+            request.context().put(HTTP_EXECUTION_STRATEGY_KEY, offloadNone());
+            return c.request(request)
+                    .flatMap(response -> {
+                        // Successful response to CONNECT never has a message body, and we are not interested in payload
+                        // body for any non-200 status code. Drain it asap to free connection and RS resources before
+                        // starting TLS handshake or propagating an error.
+                        if (response.status().statusClass() != SUCCESSFUL_2XX) {
+                            return drainPropagateError(response, new ProxyResponseException(c +
+                                    " Non-successful response from proxy CONNECT " + connectAddress, response.status()))
+                                    .shareContextOnSubscribe();
+                        }
+                        return response.messageBody().ignoreElements()
+                                .concat(handshake(c))
```

Comment:
the handshake is always added, looking for a `SslHandshakeCompletionEvent` - what happens in a non-HTTPS/TLS scenario?

Reply:
The CONNECT logic works only for TLS connections, `DefaultSingleAddressHttpClientBuilder` won't add this factory if TLS is not configured. I will add an assertion in the constructor to verify that inside `ProxyConnectLBHttpConnectionFactory` too.

Reply:
Worst case, if something is messed up, it won't find `DeferSslHandler` and will throw an exception. `ProxyConnectLBHttpConnectionFactoryTest` verifies these use-cases.

Reply:
For non-TLS (http) use-cases, requests are sent as-is to the proxy server with an absolute-form request target. Proxy will forward it to the final destination. See `HttpProxyTest` for details.

Reply:
👍 thanks!